### PR TITLE
Update to 0.16 (PocketMine API 2.1.0)

### PR DESCRIPTION
**Pull request by dktapps**
### Description

Preliminary 0.16 beta. This is up-to-date with all commits from pmmp/PocketMine-MP#42 , however further testing is needed due to Genisys vs PocketMine differences.

For details on API changes, please see pmmp/PocketMine-MP#42.

### CODE REVIEW

**Comment by dktapps** [comment]
@Muqsit Don't interfere. This is a protocol update.

**Comment by Muqsit** [comment]
Okay, sorry.

**Comment by NoHaxJustTryHqrd** [comment]
Oi ill gladly test this out rn and report bugs

**Comment by andris155** [comment]
```
29.10 19:17:45 [Server] Server thread/CRITICAL "Could not pass event 'pocketmine\event\player\PlayerPreLoginEvent' to 'EssentialsPE v1.1.3': Undefined class constant 'DATA_SHOW_NAMETAG' on EssentialsPE\EventHandlers\PlayerEvents
29.10 19:17:45 [Server] Server thread/CRITICAL Error: "Undefined class constant 'DATA_SHOW_NAMETAG'" (EXCEPTION) in "/EssentialsPE_v1.1.3.phar/src/EssentialsPE/Loader" at line 2182
```

**Comment by caspervanneck** [comment]
Floating text dont worm i test it to

**Comment by andris155** [comment]
`29.10 19:43:32 [Server] INFO Warning: assert(): Still 4677 bytes unread! failed in phar:///Genisys_.phar/src/pocketmine/network/RakLibInterface.php on line 140`

**Comment by andris155** [comment]
And missing /tp command. :D

**Comment by dktapps** [comment]
@andris155 
First issue: EssentialsPE needs updating
Second issue: beta player attempted to join your server. Turn off zend.assertions in a production environment

**Comment by flaxues** [comment]
Whenever I run Genisys 0.16 on a new world (just generated) everything works fine. But when I try to use the 0.16 branch on an older world it crashes my client. (W10 or iOS)

**Comment by erkam2002** [comment]
How can i test it  (Download?)

**Comment by DaMatrix** [comment]
you... checkout the 0.16 branch

**Comment by andris155** [comment]
Latest EssentialsPE
```
29.10 21:54:23 [Server] Server thread/INFO [EssentialsPE] Running
29.10 21:54:23 [Server] Server thread/CRITICAL "Could not pass event 'pocketmine\event\player\PlayerPreLoginEvent' to 'EssentialsPE v2.0.0': Undefined class constant 'DATA_SHOW_NAMETAG' on EssentialsPE\EventHandlers\PlayerEvents
29.10 21:54:23 [Server] Server thread/CRITICAL Error: "Undefined class constant 'DATA_SHOW_NAMETAG'" (EXCEPTION) in "/EssentialsPE_v2.0.0.phar/src/EssentialsPE/BaseFiles/BaseAPI" at line 2154
```

**Comment by andris155** [comment]
How to turn off zend.assertions? Please help.

**Comment by dktapps** [comment]
In your php.ini:

`zend.assertions=0`

**Comment by andris155** [comment]
Thanks. :)

**Comment by TheAppleGamer** [comment]
Every now and then when I relog I will respawn in the void on a vanilla server

**Comment by fuyutsuki** [comment]
>  [Server thread/CRITICAL]: ClassNotFoundException: "Class pocketmine\scheduler\CallbackTask not found" (EXCEPTION) in "/src/spl/BaseClassLoader" at line 144

CallBackTask is have been removed?

**Comment by HotFireyDeath** [comment]
I am experiencing an issue in which signs are not displayed in the world.

**Comment by Muqsit** [comment]
Putting a '#' in chat freezes your chat/

**Comment by erkam2002** [comment]
I have an bug : i have an /Hub plugin that use $this->getServer()->getDefaultLevel()->getSpawnLocation(); if i write /Hub teleported it me on spawn and if i write it again then teleport it me away xD if you want can you it test -> 88.77.9.223 port 19132

**Comment by dktapps** [comment]
@fuyutsuki This is GENISYS, NOT POCKETMINE. Take PocketMine issues to PocketMine.

For your information, CallbackTask was removed from PocketMine more than a year ago. It will also be removed from Genisys in the near future.

**Comment by erkam2002** [comment]
I get this errors Random!
Warning: assert(): Still 1784 bytes unread! failed in phar:///home/slay/Genisys.phar/src/pocketmine/network/RakLibInterface.php on line 140

Warning: assert(): Still 5089 bytes unread! failed in phar:///home/slay/Genisys.phar/src/pocketmine/network/RakLibInterface.php on line 140

Warning: assert(): Still 1783 bytes unread! failed in phar:///home/slay/Genisys.phar/src/pocketmine/network/RakLibInterface.php on line 140

**Comment by dktapps** [comment]
@erkam2002 Read my comments above about disabling assertions.

**Comment by Adam1609** [comment]
Memory leak problem while using Teleportation plugin to another world caused.

**Comment by caspervanneck** [comment]
@Adam1609 i dont have it  (it works fine)

**Comment by dktapps** [comment]
Read open issues before spamming.

**Comment by ParkChanSol** [comment]
PMMP merged. When release Genisys' 0.16?

**Comment by NoHaxJustTryHqrd** [comment]
This is not Pocketmine JEEZ

**Comment by ParkChanSol** [comment]
@NoHaxJustTryHqrd Read https://github.com/iTXTech/Genisys/issues/2126

**Comment by dktapps** [comment]
Smart people don't ask for ETAs. Now stop spamming.

**Comment by MorpheusTP** [comment]
MÖÖH

**Comment by andris155** [comment]
```
Genisys Crash Dump Sun Oct 30 12:38:42 GMT 2016

Error: Undefined class constant 'DATA_AGEABLE_FLAGS'
File: /src/pocketmine/entity/FlyingAnimal
Line: 120
Type: notice

Code:
[111]         return !$this->onGround or abs($this->motionX) > 0.00001 or abs($this->motionY) > 0.00001 or abs($this->motionZ) > 0.00001;
[112]     }
[113] 
[114]     private function generateRandomDirection(){
[115]         return new Vector3(mt_rand(-1000, 1000) / 1000, mt_rand(-500, 500) / 1000, mt_rand(-1000, 1000) / 1000);
[116]     }
[117] 
[118] 	public function initEntity(){
[119] 		parent::initEntity();
[120] 		if($this->getDataProperty(self::DATA_AGEABLE_FLAGS) === null){
[121] 			$this->setDataProperty(self::DATA_AGEABLE_FLAGS, self::DATA_TYPE_BYTE, 0);
[122] 		}
[123] 	}
[124] 
[125] 	public function isBaby(){
[126] 		return $this->getDataFlag(self::DATA_AGEABLE_FLAGS, self::DATA_FLAG_BABY);
[127] 	}
[128] 
[129]     public function attack($damage, EntityDamageEvent $source){
[130]         if($source->isCancelled()){

Backtrace:
#0 /src/pocketmine/entity/Entity(354): pocketmine\entity\Bat->initEntity(boolean)
#1 /src/pocketmine/entity/Bat(56): pocketmine\entity\Entity->__construct(pocketmine\level\format\anvil\Chunk object, pocketmine\nbt\tag\CompoundTag pocketmine\nbt\tag\CompoundTag{.pocketmine\nbt\tag\ListTag:pocketmine\nbt\tag\ListTag{.pocketmine\nbt\tag\DoubleTag:-1138.5835808632.pocketmine\nbt\tag\DoubleTag:92.394996667897.pocketmine\nbt\tag\DoubleTag:623.46979032936.}.pocketmine\nbt\tag\ListTag:pocketmine\nbt\tag\ListTag{.pocketmine\nbt\tag\DoubleTag:1.5625630121952E-6.pocketmine\nbt\tag\DoubleTag:2.0058791634436E-7.pocketmine\nbt\tag\DoubleTag:7.8886676343834E-7.}.pocketmine\nbt\tag\ListTag:pocketmine\nbt\tag\ListTag{.pocketmine\nbt\tag\FloatTag:-63.212871551514.pocketmine\nbt\tag\FloatTag:-83.462692260742.}.pocketmine\nbt\tag\ByteTag:0.pocketmine\nbt\tag\FloatTag:101.4143371582.pocketmine\nbt\tag\ShortTag:0.pocketmine\nbt\tag\ShortTag:1200.pocketmine\nbt\tag\ByteTag:0.pocketmine\nbt\tag\ByteTag:0.pocketmine\nbt\tag\ShortTag:6.pocketmine\nbt\tag\StringTag:Bat.})
#2 /src/pocketmine/entity/Entity(556): pocketmine\entity\Bat->__construct(pocketmine\level\format\anvil\Chunk object, pocketmine\nbt\tag\CompoundTag pocketmine\nbt\tag\CompoundTag{.pocketmine\nbt\tag\ListTag:pocketmine\nbt\tag\ListTag{.pocketmine\nbt\tag\DoubleTag:-1138.5835808632.pocketmine\nbt\tag\DoubleTag:92.394996667897.pocketmine\nbt\tag\DoubleTag:623.46979032936.}.pocketmine\nbt\tag\ListTag:pocketmine\nbt\tag\ListTag{.pocketmine\nbt\tag\DoubleTag:1.5625630121952E-6.pocketmine\nbt\tag\DoubleTag:2.0058791634436E-7.pocketmine\nbt\tag\DoubleTag:7.8886676343834E-7.}.pocketmine\nbt\tag\ListTag:pocketmine\nbt\tag\ListTag{.pocketmine\nbt\tag\FloatTag:-63.212871551514.pocketmine\nbt\tag\FloatTag:-83.462692260742.}.pocketmine\nbt\tag\ByteTag:0.pocketmine\nbt\tag\FloatTag:101.4143371582.pocketmine\nbt\tag\ShortTag:0.pocketmine\nbt\tag\ShortTag:1200.pocketmine\nbt\tag\ByteTag:0.pocketmine\nbt\tag\ByteTag:0.pocketmine\nbt\tag\ShortTag:6.pocketmine\nbt\tag\StringTag:Bat.})
#3 /src/pocketmine/level/format/generic/BaseFullChunk(148): pocketmine\entity\Entity::createEntity(string Bat, pocketmine\level\format\anvil\Chunk object, pocketmine\nbt\tag\CompoundTag pocketmine\nbt\tag\CompoundTag{.pocketmine\nbt\tag\ListTag:pocketmine\nbt\tag\ListTag{.pocketmine\nbt\tag\DoubleTag:-1138.5835808632.pocketmine\nbt\tag\DoubleTag:92.394996667897.pocketmine\nbt\tag\DoubleTag:623.46979032936.}.pocketmine\nbt\tag\ListTag:pocketmine\nbt\tag\ListTag{.pocketmine\nbt\tag\DoubleTag:1.5625630121952E-6.pocketmine\nbt\tag\DoubleTag:2.0058791634436E-7.pocketmine\nbt\tag\DoubleTag:7.8886676343834E-7.}.pocketmine\nbt\tag\ListTag:pocketmine\nbt\tag\ListTag{.pocketmine\nbt\tag\FloatTag:-63.212871551514.pocketmine\nbt\tag\FloatTag:-83.462692260742.}.pocketmine\nbt\tag\ByteTag:0.pocketmine\nbt\tag\FloatTag:101.4143371582.pocketmine\nbt\tag\ShortTag:0.pocketmine\nbt\tag\ShortTag:1200.pocketmine\nbt\tag\ByteTag:0.pocketmine\nbt\tag\ByteTag:0.pocketmine\nbt\tag\ShortTag:6.pocketmine\nbt\tag\StringTag:Bat.})
#4 /src/pocketmine/level/Level(2794): pocketmine\level\format\generic\BaseFullChunk->initChunk(boolean)
#5 /src/pocketmine/level/Level(696): pocketmine\level\Level->loadChunk(integer -72, integer 38)
#6 /src/pocketmine/Player(957): pocketmine\level\Level->registerChunkLoader(pocketmine\Player Player(342), integer -72, integer 38, boolean 1)
#7 /src/pocketmine/Player(1950): pocketmine\Player->sendNextChunk(boolean)
#8 /src/pocketmine/Server(2853): pocketmine\Player->checkNetwork(boolean)
#9 /src/pocketmine/Server(2542): pocketmine\Server->tick(boolean)
#10 /src/pocketmine/Server(2420): pocketmine\Server->tickProcessor(boolean)
#11 /src/pocketmine/Server(2010): pocketmine\Server->start(boolean)
#12 /src/pocketmine/PocketMine(476): pocketmine\Server->__construct(pocketmine\CompatibleClassLoader object, pocketmine\utils\MainLogger object, string phar:///home/minecraft/multicraft/servers/server2/Genisys_.phar/, string /home/minecraft/multicraft/servers/server2/, string /home/minecraft/multicraft/servers/server2/plugins/, string unknown)
#13 (1): require_once(string phar:///home/minecraft/multicraft/servers/server2/Genisys_.phar/src/pocketmine/PocketMine.php)

Genisys version: 0000000000000000000000000000000000000000 [Protocol 91; API 2.1.0]
uname -a: Linux server.theoldcrafters.hol.es 3.13.0-98-generic #145-Ubuntu SMP Sat Oct 8 20:13:07 UTC 2016 x86_64
PHP version: 7.0.12
Zend version: 3.0.0
OS : Linux, linux

Server uptime: 46 minutes 29 seconds
Number of loaded worlds: 3
Players online: 10/80

Loaded plugins:
AntiSpam 1.3 by xBeastMode for API(s) 2.0.0
CombatLogger 1.2.0 by Lambo for API(s) 1.12.0
NoAdvertisingPE 1.0.1 by Vaivez66 for API(s) 1.12.0
MineMarket 0.1.0 by  for API(s) 1.0.0
MagicWE 0.2.7 by thebigsmileXD for API(s) 1.13.0, 2.0.0
EconomyAPI 2.0.9 by onebone for API(s) 1.0.0, 1.1.0, 1.2.0, 1.2.1, 1.3.0, 1.3.1, 1.4.0, 1.4.1
MagicTelePortal 1.3.2 by aliuly for API(s) 1.10.0
TrewManager 1.0 by HypertextPP for API(s) 2.0.0
AntiBotPE 1.1 by luca28pet for API(s) 2.0.0
SimpleMessages 1.2 by Dutok for API(s) 1.0.0
LimitedCreative 1.0.0 by basprohop for API(s) 1.10.0
ServerAuth 2.13 by EvolSoft for API(s) 1.14.0
NetherChests 1.1.1 by aliuly for API(s) 1.12.0
iProtector 3.1b by LDX for API(s) 1.9.0, 2.0.0
ServerKits 1.3 by EvolSoft for API(s) 1.11.0
EssentialsPE 1.1.3 by LegendsOfMCPE Team for API(s) 1.12.0
EconomyShop 2.0.7 by onebone for API(s) 1.12.0
BanItem 2.2 by LDX for API(s) 1.0.0, 2.0.0
KillMoneyFix 1 by CrazedMiner for API(s) 1.8.0
PurePerms 1.3.7 by 64FF00 & ProjectInfinity for API(s) 1.13.0, 2.0.0
ClearLagg 1.1.0 by LegendOfMCPE for API(s) 1.0.0
PureChat 1.4.0 by 64FF00 for API(s) 1.13.0, 2.0.0
```

**Comment by TheAppleGamer** [comment]
Enchants don't work anymore

**Comment by TheAppleGamer** [comment]
Not sure if it's FNPC or the 0.16 but leads come from entity nametags

**Comment by dktapps** [comment]
@TheAppleGamer yes, that's FNPC, the metadata for leads changed. Enchanting is not currently supported due to changes we haven't been able to work out yet.

**Comment by TheAppleGamer** [comment]
@dktapps Alright on both. I'll let the person who opened a PR on FNPC for 0.16 know that they failed lol

**Comment by Muqsit** [comment]
@dktapps Crashes occur while spawning particles. Particle numbers need an update. Stead****2 has it.

**Comment by fuyutsuki** [comment]
@dktapps
Oh, Ok…
Is there an alternative to CallbackTask?

**Comment by dktapps** [comment]
@fuyutsuki You should use PluginTask instead.

**Comment by dktapps** [comment]
@Muqsit I'm aware... these are from the disassembly: https://github.com/pmmp/PocketMine-MP/commit/108267f16728baaa2d0cc826c2ce9deeec0da2ec

Untested however, have a play with them if you like.

**Comment by Tarik02** [comment]
@dktapps,
```
[18:26:27] [Server thread/CRITICAL]: "Could not pass event 'Tarik02\leveldata\LevelDataLoadedEvent' to 'Hub v1.8.0': Return value of pocketmine\entity\Entity::setImmobile() must be of the type boolean, null returned on Tarik02\Hub\HubPlugin
[18:26:27] [Server thread/CRITICAL]: TypeError: "Return value of pocketmine\entity\Entity::setImmobile() must be of the type boolean, null returned" (EXCEPTION) in "/src/pocketmine/entity/Entity" at line 444
```

**Comment by dktapps** [comment]
@Tarik02 Fixed

**Comment by Tarik02** [comment]
@dktapps Thanks.

**Comment by andris155** [comment]
```
30.10 17:58:18 [Server] Server thread/CRITICAL Unhandled exception executing command 'warp arena' in warp: Argument 1 passed to pocketmine\Player::teleport() must be an instance of pocketmine\math\Vector3, array given, called in phar:///plugins/EssentialsPE_v1.1.3.phar/src/EssentialsPE/Commands/Warp/Warp.php on line 52
30.10 17:58:18 [Server] Server thread/CRITICAL TypeError: "Argument 1 passed to pocketmine\Player::teleport() must be an instance of pocketmine\math\Vector3, array given, called in phar:///plugins/EssentialsPE_v1.1.3.phar/src/EssentialsPE/Commands/Warp/Warp.php on line 52" (EXCEPTION) in "/src/pocketmine/Player" at line 4142
```

**Comment by dktapps** [comment]
@andris155 Please report that to EssentialsPE's issue tracker.

**Comment by andris155** [comment]
0.15.10 Genisys - works perfectly EssentialsPE.
0.16.0(.5) - not work perfectly EssentialsPE. 
The EssentialsPE author not update and not read issues.

**Comment by andris155** [comment]
https://github.com/LegendOfMCPE/EssentialsPE/pull/295

**Comment by NoHaxJustTryHqrd** [comment]
Possibility of updating the synapse api to 0.16? just saying...

**Comment by Muqsit** [comment]
BossEventPacket?

**Comment by fuyutsuki** [comment]
@dktapps Ok, thank you!

**Comment by NoHaxJustTryHqrd** [comment]
@dktapps synpase api in genisys needs updating. Just sayinf

**Comment by orlando092** [comment]
Players cant see chest and signs

**Comment by BlawoHD** [comment]
0.16.0 is closed?

**Comment by dktapps** [comment]
Fixed an issue with NBT in MCRegion worlds with async-chunk-request enabled. Please test and report back whether it fixes the issues with invisible tiles.

**Comment by Muqsit** [comment]
Works fine. Btw, $creeperentity->setPowered(true) is broken.

**Comment by dktapps** [comment]
@Muqsit Entity metadata changes, I'll fix it some other time. Right now all I'm interested in is stabilizing this so it's actually usable. API can wait.

**Comment by Muqsit** [comment]
Makes sense

**Comment by ImagicalGamer** [comment]
Entity problems but everything else works fine.

**Comment by TheAppleGamer** [comment]
When multiple players tp to random areas that have already generated, in the sky, are unable to move or do anything but chat until they relog.

**Comment by TheAppleGamer** [comment]
When multiple players tp to random areas that have already generated, in the sky, are unable to move or do anything but chat until they relog.

**Comment by SuperAlecs** [comment]
the primed tnt is buggy @dktapps

**Comment by kabluinc** [comment]
After doing /kill, player becomes glitched and after reclogging, spawns slightly further away from original spawn and gets extreme lag. this can be seen at the server kidzzone.ca immediately after u join it

**Comment by dktapps** [comment]
@kabluinc I'm aware of the issue, actually working on it right now. Stay tuned :)

**Comment by HasanBrandsXD** [comment]
Where can I get the Genisys 0.16?

**Comment by kabluinc** [comment]
Probably not an important issue but minecarts seem to be floating in the air when placed

**Comment by DaMatrix** [comment]
@HasanBrandsXD 
screenshot from 2016-11-01 18-23-20

**Comment by dktapps** [comment]
If only I could pin GitHub issue comments...

**Comment by TheAppleGamer** [comment]
getName() on null

**Comment by TheAppleGamer** [comment]
Warning: Illegal offset type in isset or empty in C:\Users\drewb\Desktop\Gensisys\src\pocketmine\entity\Entity.php on line 483

**Comment by andris155** [comment]
```
2016-11-01 [10:52:14] [Server thread/INFO]: [+] Pvpgirl116
2016-11-01 [10:52:20] [Server thread/INFO]:  lost connection: Login timeout
2016-11-01 [10:52:25] [Server thread/CRITICAL]: InvalidArgumentException: "Must have exactly 16 bytes" (EXCEPTION) in "/src/pocketmine/utils/UUID" at line 66
2016-11-01 [10:52:35] [Server thread/INFO]:  lost connection: Login timeout
2016-11-01 [10:52:35] [Server thread/INFO]: [-] Pvpgirl116
2016-11-01 [10:52:35] [Server thread/INFO]: Pvpgirl116 lost connection: client disconnect
2016-11-01 [10:52:58] [Server thread/CRITICAL]: InvalidArgumentException: "Must have exactly 16 bytes" (EXCEPTION) in "/src/pocketmine/utils/UUID" at line 66
```

**Comment by inferboy** [comment]
When will it get accepted?

**Comment by dktapps** [comment]
When it doesn't crash anymore, I'll merge it.

**Comment by inferboy** [comment]
Hahahha

File: src/pocketmine/Player.php [hunk 1]
```diff
@@ -2266,7 +2266,7 @@ public function handleDataPacket(DataPacket $packet){
 					break;
 				}
 
-				if(!in_array($packet->protocol, ProtocolInfo::ACCEPTED_PROTOCOLS)){
+				if($packet->protocol !== ProtocolInfo::CURRENT_PROTOCOL){
```
Comment (DaMatrix) [inline src/pocketmine/Player.php]: really? what if there's another one of those updates that changes the protocol number but not the protocol, like 0.15.9 to 0.15.10? all that did was add a texture pack.

Comment (dktapps) [inline src/pocketmine/Player.php]: ... to the best of your/our knowledge. Until we can say for 100% certain what protocol changes were made and what behaviours will result, we should not mix them. MCPE doesn't allow mixing them, and there's a damn good reason for that.

**Comment by Muqsit** [comment]
So the main causes of crashes right now are:
- 0.16
- 0.16
- particle protocols
- 0.16

**Comment by dktapps** [comment]
Still testing particles, some of their data has changed. The crashy one (12) is ItemBreakParticle, it crashes you because it's item break of air. Give it data and it's happy.

**Comment by Shadow166050** [comment]
Signs Are Randomly Replaced With Chest, Thinking Its a Tile Issue, I Don't See Any Errors In Console

**Comment by TheAppleGamer** [comment]
Unless there is a new way to cancel natural regen in plugins, I would say that's broken.

**Comment by inferboy** [comment]
What happened to getIteminHand()?

**Comment by Tarik02** [comment]
When I set or add experience, it not shows in client. I can see it only after open and close inventory or any other menu.

**Comment by dktapps** [comment]
@Tarik02 Blame Mojang

**Comment by Muqsit** [comment]
Wait...this bug doesn't happen with me.

**Comment by inferboy** [comment]
Omg fix and accept changes already

**Comment by inferboy** [comment]
Hope u accept it before they release 0.17.0

**Comment by TheAppleGamer** [comment]
@inferboy If you think it's that quick and easy to make a compatible server software with little bugs then you thought wrong.

**Comment by dktapps** [comment]
This PR is here for the purposes of testing. When it is stable I will merge it. Currently the bug reports are still coming in torrents, so it remains here.

**Comment by TheAppleGamer** [comment]
@dktapps It appears that effects cause an error in console but I removed a semi important line and no more errors

**Comment by flaxues** [comment]
Yeah update already it's really simple just bump the protocol and everything is fixed -_- Thats not how it fucking works ok? Be happy that @dktapps Is helping us all!

**Comment by inferboy** [comment]
Yea be happy @flaxues that hes helping, @flaxues is such a noob that he goes around asking for factions plugin lmao grow up and learn php

**Comment by dktapps** [comment]
Enough already, take your arguments somewhere else thanks.

**Comment by inferboy** [comment]
Great work @dktapps You're the best!

**Comment by DaMatrix** [comment]
WOOOOOO!!! HE MERGED IT!!!

**Comment by dktapps** [comment]
It took me a while to isolate a very irritating and frequent crash on player join, it should now be fixed. There's still plenty of mess to tidy up, but it should be just about usable now.

**Comment by KairusDarkSeeker** [comment]
Seems no bugs for me :D

**Pull request merged**